osd,librados: cmpext support

ddiss opened the pull request:
This is a follow up to #8175 and #8568. The patch-set carries @mikechristie 's implementation of OSD cmpext support.
The librados API, based on @mikechristie 's initial C++ work, has been extended to handle transparent miscompare response unmarshalling. AIO support and a C API has also been added, along with corresponding unit tests.
Some aspects of the patchset that could potentially be improved:
- The C API invokes an unnecessary memcpy() on miscompare
  - this was discussed upstream via http://www.spinics.net/lists/ceph-devel/msg29969.html
- The IoCtxImpl::aio_cmpext() helpers used by the AIO cmpext APIs are a little convoluted
  - prepare_cmpext_op is used for C++. m_ops.cmpext() + prepare_read_op() for C

liewegas commented:
@jdurgin ?

yuyuyu101 commented on src/include/rados/librados.hpp:
```diff
@@ -907,6 +911,37 @@ namespace librados
+     * @param c what to do when the read is complete
+     * @param mismatch_bl buffer filled with object data on mismatch
+     * @param mismatch_off offset into buffer where the mismatch was detected
+     * @param off the offset to start comparison from in the object
```
lack of cmp_bl

ddiss replied:
Will fix next round, thanks!

yuyuyu101 commented on src/include/rados/librados.hpp:
```diff
@@ -907,6 +911,37 @@ namespace librados
+		   bufferlist *mismatch_bl,
+		   uint64_t *mismatch_off,
+		   uint64_t off,
+		   bufferlist& cmp_bl, uint64_t snapid);
```
snapid?

yuyuyu101 commented on src/librados/IoCtxImpl.cc:
```diff
@@ -868,6 +868,57 @@ int librados::IoCtxImpl::aio_sparse_read(const object_t oid,
+		      mismatch_off, NULL);
+
+  Objecter::Op *o = objecter->prepare_read_op(
+    oid, oloc, onack->m_ops, snapid, NULL, 0, onack, &c->objver);
```
why do we need to pass in snapid? snap_seq instead?

ddiss replied:
This differs across librados::IoCtxImpl::aio_sparse_read() and librados::IoCtxImpl::aio_read() - both take a snapid parameter, aio_sparse_read() ignores it and uses snap_seq (i.e. the set_snap_read() value), whereas aio_read() uses snapid.
I don't know which variant is correct here.

ddiss replied:
FWIW, in case aio_sparse_read() is incorrectly ignoring the snapid parameter, I've pushed: https://github.com/ddiss/ceph/commits/librados_sparse_read_use_snapid
Will wait on feedback before submitting.

jdurgin replied:
that patch is right, IoCtxImpl::aio_sparse_read should be using the snapid parameter. That parameter was added to aio_read and aio_sparse_read to support passing it through new versions of those functions in the api - you can see the original versions just pass in io_ctx_impl->snap_seq.

I think it's fine to omit the snap_id parameter for all these cmpext functions, and just use snap_seq. If folks care, they can use the generalized ObjectOperation version which allows passing a snap_id via the operate() methods.

ddiss replied:
I've submitted the sparse-read read snapid fix via:
https://github.com/ceph/ceph/pull/9116

I'll drop the cmpext snap_id parameters and push a new round today.

Thanks for the feedback Josh!

ddiss commented:
Changes since previous version:
- clean-up doxygen comments in header files
- rebase against master
- drop changelog from Mike's OSD op commit message

Feedback appreciated.

ddiss commented:
One further question... can/should I add a new feature flag for the cmpext+writesame operations, or is it expected that they'll just be covered by a generic flag for the next release?

ddiss commented:
Anyone get a chance to take a look here? @jdurgin ?

jdurgin commented:
not it depth yet, will look carefully today. Generally we don't use feature flags for new ops. since new versions of clients often need new server side cls ops, we always recommend upgrading osds first and not using new features until that's done. 

On the client side for krbd we can log the error if we get enosys/enotsupp or whatever it is from an old osd.

jdurgin commented:
looks good to me

ddiss commented:
Changes since previous version:
- removed cmpext snapid parameters
  + squashed patch at https://github.com/ddiss/ceph/commit/03bd0c340e099e620014153f49ddeb42c6bf6a3b
- rebased against latest master

jdurgin commented:
thanks, changes lgtm

yuriw commented:
This PR seems like broke many jobs, see http://pulpito.ceph.com/yuriw-2016-05-16_13:50:33-rados-wip-yuri-testing2-distro-basic-smithi/

athanatos commented:
sjust@teuthology:/a/yuriw-2016-05-16_13:50:33-rados-wip-yuri-testing2-distro-basic-smithi/195939

appears to be one example

2016-05-16T14:07:45.265 INFO:tasks.rados.rados.0.smithi040.stdout:1: write initial oid 1 oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
2016-05-16T14:07:45.265 INFO:tasks.rados.rados.0.smithi040.stdout:1:  seq_num 0 ranges {416807=475249,1542129=474678}
2016-05-16T14:07:45.276 INFO:tasks.rados.rados.0.smithi040.stdout:1:  writing smithi04017127-1 oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo from 416807 to 892056 tid 1
2016-05-16T14:07:45.283 INFO:tasks.rados.rados.0.smithi040.stdout:1:  writing smithi04017127-1 oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo from 1542129 to 2016807 tid 2
2016-05-16T14:07:45.283 INFO:tasks.rados.rados.0.smithi040.stdout:2: write initial oid 2
2016-05-16T14:07:45.284 INFO:tasks.rados.rados.0.smithi040.stdout:2:  seq_num 1 ranges {416807=475249,1542129=474678}
2016-05-16T14:07:45.284 INFO:tasks.rados.rados.0.smithi040.stderr:*** Error in `ceph_test_rados': free(): invalid pointer: 0x00007f2bf96287b8 ***
2016-05-16T14:07:45.309 INFO:teuthology.orchestra.run.smithi050:Running: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 30 ceph --cluster ceph --admin-daemon /v

jdurgin commented:
Guessing https://github.com/ceph/ceph/pull/8911/commits/7ebd83b85075b21933cb15a4eac62ed787b690ca is causing a misuse of memory

ddiss commented:
Sorry, I didn't see this in local testing, will take a look and report back. Thanks!

ddiss commented:
I haven't forgotten about this, and will revisit the ceph_test_rados failure when I get a spare moment.

Nefelim4ag commented:
@ddiss, Ping, any updates?

Thanks.
P.S. I'm just waiting when rbd iSCSI will be ready

dachary commented:
needs a rebase

ddiss commented:
Yes, this needs a rebase, in addition to the fix for the ceph_test_rados failure. I've kept it open for tracking, but will close it (and resubmit later) if it's getting in the way.

liuy commented:
@ddiss Hi, We (from China Mobile) are going to help with rbd iscsi and would like to work on this and continue the code to walk this compext support into ceph.

ddiss commented:
I've rebased this patch-set against current master and pushed to my osd-librados-cmpext_v3 branch:
https://github.com/ddiss/ceph/tree/osd-librados-cmpext_v3

Will submit from there, once I have it cleaned up and passing passing all tests.

ddiss commented:
With the rebased patch-set (https://github.com/ddiss/ceph/tree/osd-librados-cmpext_v3) I'm no longer able to hit the ceph_test_rados failures seen earlier.
The only modifications I've made, aside from the rebase, is a fix for the rados_aio_cmpext LTTNG tracepoints.

mikechristie commented:
Hey @ddiss did your last comment on Jan 5th mean it is all working now and is it ready for submission?

Or, do you need any help at all? We have people that want to help test and/or code whatever else is needed.

dillaman commented:
@ddiss ping -- is your branch ready for a new PR?

ddiss commented:
@dillaman : yes I think so, given that I'm no longer seeing the earlier ceph_test_rados regression. Reopening here - would appreciate any extra tests / review others could throw at this branch.
